File: python-sdk/expansion/interfaces/rpc/passive_scanner.py
from typing import Optional, NamedTuple, List, Tuple
from enum import Enum

import expansion.protocol as protocol
import expansion.protocol.Protocol_pb2 as public
from expansion.transport import IOTerminal, Channel
import expansion.utils as utils
import expansion.types as types


class Specification(NamedTuple):
    scanning_radius_km: int
    max_update_time_ms: int


class PassiveScannerI(IOTerminal):

    class Status(Enum):
        SUCCESS = "success"
        MONITORING_FAILED = "monitoring failed"

        # Internal SDK statuses:
        FAILED_TO_SEND_REQUEST = "failed to send request"
        RESPONSE_TIMEOUT = "response timeout"
        UNEXPECTED_RESPONSE = "unexpected response"
        CHANNEL_CLOSED = "channel closed"
        CANCELED = "operation canceled"

        def is_success(self):
            return self == PassiveScannerI.Status.SUCCESS

        def is_timeout(self):
            return self == PassiveScannerI.Status.RESPONSE_TIMEOUT

    def __init__(self, name: Optional[str] = None):
        super().__init__(name=name or utils.generate_name(PassiveScannerI))
        self.specification: Optional[Specification] = None

    @Channel.return_on_close(None)
    async def get_specification(self, timeout: float = 0.5)\
            -> Optional[Specification]:
        request = public.Message()
        request.passive_scanner.specification_req = True
        if not self.send(message=request):
            return None
        response, _ = await self.wait_message(timeout=timeout)
        if not response:
            return None
        spec = protocol.get_message_field(response, ["passive_scanner", "specification"])
        if not spec:
            return None
        return Specification(scanning_radius_km=spec.scanning_radius_km,
                             max_update_time_ms=spec.max_update_time_ms)

    @Channel.return_on_close(Status.CHANNEL_CLOSED)
    def start_monitoring(self, timeout: float = 0.5) -> Status:
        request = public.Message()
        request.passive_scanner.monitor = True
        response, _ = await self.wait_message(timeout=timeout)
        if not response:
            return PassiveScannerI.Status.RESPONSE_TIMEOUT
        ack = protocol.get_message_field(response,
                                         ["passive_scanner", "monitor_ack"])
        return PassiveScannerI.Status.SUCCESS \
            if ack else PassiveScannerI.Status.MONITORING_FAILED

    @Channel.return_on_close((Status.MONITORING_FAILED, list()))
    def wait_update(self, timeout: float = 1) \
            -> Tuple[Status, List[types.PhysicalObject]]:
        response = await self.wait_message(timeout=timeout)
        if not response:
            return PassiveScannerI.Status.SUCCESS, list()
        update = protocol.get_message_field(
            response, ["passive_scanner", "update"])
        if not update:
            return PassiveScannerI.Status.UNEXPECTED_RESPONSE, list()

        return PassiveScannerI.Status.SUCCESS, [
            types.PhysicalObject(
                object_type=types.ObjectType.from_protobuf(item.object_type),
                object_id=item.id,
                position=types.Position(
                    x=item.x,
                    y=item.y,
                    velocity=types.Vector(x=item.vx, y=item.vy)
                ),
                radius=item.r
            )
            for item in update.items
        ]

File: typescript-sdk/lowlevel/passive_scanner.ts
import { create } from "@bufbuild/protobuf";
import * as msg from "../Protocol_pb.js";
import * as types from "../types/index.js";
import { Session } from "./session.js";


export type PassiveScannerSpecification = {
    scanning_radius_km: number;
    max_update_time_ms: number;
}

export class PassiveScanner {

    constructor(private session: Session) {}

    async send_specification_request(): Promise<types.Status> {
        const request = create(msg.IPassiveScannerSchema, {
            choice: { case: "specificationReq", value: true },
        });
        return this.send(request);
    }

    async wait_specification(timeout: number = 500)
    : Promise<[types.Status, PassiveScannerSpecification | undefined]>
    {
        const [status, response] = await this.wait(timeout);
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined];
        }
        if (response.choice.case != "specification") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined];
        }
        const spec = response.choice.value;
        return [types.Status.ok(), {
            scanning_radius_km: spec.scanningRadiusKm,
            max_update_time_ms: spec.maxUpdateTimeMs,
        }];
    }

    async send_monitor_request(): Promise<types.Status> {
        const request = create(msg.IPassiveScannerSchema, {
            choice: { case: "monitor", value: true },
        });
        return this.send(request);
    }

    async wait_monitor_ack(timeout: number = 500)
    : Promise<[types.Status, boolean | undefined]>
    {
        const [status, response] = await this.wait(timeout);
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined];
        }
        if (response.choice.case != "monitorAck") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined];
        }
        return [types.Status.ok(), response.choice.value];
    }

    async wait_update(timeout: number = 500)
    : Promise<[types.Status, types.PhysicalObject[] | undefined]>
    {
        const [status, response, timestamp] = await this.wait(timeout);
        if (status.is_timeout()) {
            return [status, undefined];
        }
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined];
        }
        if (response.choice.case != "update") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined];
        }
        const objects = (response.choice.value.items ?? []).map(
            (item) => types.physicalObjectFromProtobuf(item, timestamp));
        return [types.Status.ok(), objects];
    }

    private async send(request: msg.IPassiveScanner): Promise<types.Status> {
        const message = create(msg.MessageSchema, {
            choice: { case: "passiveScanner", value: request },
        });
        return this.session.send(message);
    }

    private async wait(timeout_ms: number = 500)
    : Promise<[types.Status, msg.IPassiveScanner | undefined, bigint]>
    {
        const [status, response] = await this.session.wait(timeout_ms);
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined, BigInt(0)];
        }
        if (response.choice.case != "passiveScanner") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined, BigInt(0)];
        }
        return [types.Status.ok(), response.choice.value, types.asUint64(response.timestamp)];
    }

}
